Embedded tracker: Return external IPs to a peer with external IP

SeigneurSerpent opened the pull request:
A tracker behind NAT serving a private torrent should provide suitable IPs
o peers

-Each peer has announcement and connection IP addresses. When announcement request comes from a public client to a tracker behind NAT serving a private torrent, the tracker should provide public IP in response, selecting between connection and announcement IPs. It allows the client to start downloading immediately without waiting for other peers to connect.
-Set announce_ip to external ip obtained during external_ip_alert handler
-Handle url parameter decoding by the tracker itself. Bittorrent tracker protocol uses RFC1738 to percent-decode parameters, which is based on decoding single bytes. Qt5 QUrlQuery decoding function has difficulty decoding sequences 0x80-0xFF.

ngosang commented:
This feature is necessary but I've looked at the code and the implementation seems a little complicated. Maybe @glassez  can help you simplify things.

SeigneurSerpent commented:
Let me elaborate on the changes introduced to clear a few things up.

`session.cpp` is modified to set external IP address to `announce_ip`. That is how a client tells its public IP even operating in a private network.

Announcement URL parameters `info_hash` and `peer_id` are just random 20-byte binary strings. Qt5's `QUrlQuery` wouldn't decode the parameters fully as it normally expects UTF-8 strings. Sometimes it could lead to different "half-decoded" strings corresponding to one binary string, making the decoding function non-bijective (not one-to-one). Therefore files `webapplication.cpp`, `webapplication.h`, `requestparser.h`, `requestparser.cpp`, `irequesthandler.h` and `connection.cpp` are changed to make url parameters decoding optional and turn it off for embedded tracker (via `configureRequestParser()` request and then invoking `disableURLParameterDecoding()` on `parser `), and let the tracker percent-decode the parameters. Still, when string decoding fails for some reason, the string is stored in a hashtable as a key thereby preserving the way tracker used to operate before for erroneous strings. That’s why several `if` clauses are introduced in `tracker.cpp`

Static functions in `IpUtil` and `isNonRoutable()` in particular are introduced in `tracker.cpp` to determine whether given IP address is public (external). These functions are used in `Peer::getIpForClient(clientIp)` to retrieve public IP for a given peer if the `clientIp` is public. Otherwise it falls back to connection IP.

If you have any questions or suggestions for improvement, please don't hesitate to tell.

glassez commented:
>Qt5's  QUrlQuery  wouldn't decode the parameters fully as it normally expects UTF-8 strings. Sometimes it could lead to different "half-decoded" strings corresponding to one binary string, making the decoding function non-bijective (not one-to-one). 

For a start, please describe this problem in detail. Give me an example URL that is decoded incorrectly, so I can check it out.

SeigneurSerpent commented:
@glassez
Lets consider simple test that decodes sequences "()A%7BB" and "%28%29A%7BB". These sequences represent the same binary string. The default QUrl::PrettyDecoded doesn't do much and produces ambiguity. QUrl::FullyDecoded seems to decode every character, but replaces non-UTF8 0x80-0xFF characters with a predefined character. It is evident when comparing decoded  strings "%80%00" and "%FF%00".

Reference seems to corroborate this:
http://doc.qt.io/qt-5/qurl.html#full-decoding
"Non-UTF-8 sequences: URLs may contain sequences of percent-encoded characters that do not form valid UTF-8 sequences. Since URLs need to be decoded using UTF-8, any decoder failure will result in the QString containing one or more replacement characters where the sequence existed."

```
#include "libtorrent/escape_string.hpp"
#include <string>
#include <QByteArray>
#include <QUrlQuery>

QString decode(const QString& encoded, QUrl::ComponentFormattingOptions encoding = QUrl::PrettyDecoded)
{
    QString url = QString("announce?a=") + encoded;
    QListIterator<QPair<QString, QString> > i(QUrlQuery(url).queryItems(encoding));

    QString decoded = i.next().second;

    qDebug() << "Decoded: " << decoded;
    qDebug() << "Hex: " << decoded.toLocal8Bit().toHex();

    return decoded;
}

void test()
{
    char s1[] = "()A%7BB";
    // perfectly valid percent-encoded string that represents the same binary string as s1, but characters '(' and ')' are encoded
    char s2[] = "%28%29A%7BB";
    boost::system::error_code ec1, ec2;
    qDebug() << "Libtorrent:";
    qDebug() << "Same binary? " << (libtorrent::unescape_string(s1, ec1) == libtorrent::unescape_string(s2, ec2));
    qDebug() << "";

    QString e1(s1);
    QString e2(s2);
    qDebug() << "default(PrettyDecoded):";
    qDebug() << "Same binary? " << (decode(e1) == decode(e2));
    qDebug() << "";

    // FullyDecoded seems to help decode symbols, but...
    qDebug() << "FullyDecoded:";
    qDebug() << "Same binary? " << (decode(e1, QUrl::FullyDecoded) == decode(e2, QUrl::FullyDecoded));
    qDebug() << "";

    // non-UTF8 sequences are replaced with a predefined charachter
    qDebug() << "FullyDecoded with Non-UTF8 characters:";
    QString c1("%80%00");
    QString c2("%FF%00");
    qDebug() << "Same binary? " << (decode(c1, QUrl::FullyDecoded) == decode(c2, QUrl::FullyDecoded));
    qDebug() << "";
}
```

```
Libtorrent:
Same binary?  true

default(PrettyDecoded):
Decoded:  "%28%29A{B"
Hex:  "253238253239417b42"
Decoded:  "()A{B"
Hex:  "2829417b42"
Same binary?  false

FullyDecoded:
Decoded:  "()A{B"
Hex:  "2829417b42"
Decoded:  "()A{B"
Hex:  "2829417b42"
Same binary?  true

FullyDecoded with Non-UTF8 characters:
Decoded:  "?\u0000"
Hex:  "3f00"
Decoded:  "?\u0000"
Hex:  "3f00"
Same binary?  true
```

glassez commented:
@SeigneurSerpent Since there are two independent problems, it would be better if you will divide this PR into two: one for IP handling and the other for URL decoding. This may speed up reviewing and merging. (leave this PR for the URL decoding issue, because there are associated comments)

glassez commented:
>Lets consider a simple test that decodes sequences "()A%7BB" and "%28%29A%7BB". These sequences represent the same binary string. The default QUrl::PrettyDecoded doesn't do much and produces ambiguity. QUrl::FullyDecoded seems to decode every character, but replaces non-UTF8 0x80-0xFF characters with a predefined character. It is evident when comparing decoded strings "%80%00" and "%FF%00".

OK. That's all clear. But it's too abstract example.
What about actual cases? What parameters can contain such escapes, under what conditions, and what clients can produce this? And, most importantly, can we consider such requests valid from the BitTorrent protocol point of view?

SeigneurSerpent commented:
@glassez 
> Since there are two independent problems, it would be better if you will divide this PR into two: one for IP handling and the other for URL decoding. This may speed up reviewing and merging. (leave this PR for the URL decoding issue, because there are associated comments)

The code is difficult to divide seamlessly into two separate updates, because both rely on `Peer` class interface, which has been modified to incorporate both changes. Therefore the separation would drag changes from both updates and lead to multiple merge conflicts. Maybe it is better to move some IP-related code to a separate file to avoid clutter. Another option is to first merge ulr-decoding PR and then create remote ip handling PR.

> What about actual cases? What parameters can contain such escapes, under what conditions, and what clients can produce this? And, most importantly, can we consider such requests valid from the BitTorrent protocol point of view?

According to the specification, info_hash and peer_id are 20-byte binary strings that must be percent-escaped:
[https://wiki.theory.org/BitTorrentSpecification](url)

> Note that all binary data in the URL (particularly info_hash and peer_id) must be properly escaped.

> info_hash: urlencoded 20-byte SHA1 hash of the value of the info key from the Metainfo file.

> peer_id: urlencoded 20-byte string used as a unique ID for the client, generated by the client at startup.

The following page describes bittorrent tracker protocol in detail:
[https://wiki.theory.org/BitTorrent_Tracker_Protocol](url)

glassez commented:
>Maybe it is better to move some IP-related code to a separate file to avoid clutter.

Do not rush to do it. Let me watch current implementation.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -56,22 +61,156 @@ bool Peer::operator==(const Peer &other) const
+{
+    boost::system::error_code ec;
+
+    QByteArray qtdecoded = QByteArray::fromPercentEncoding(escaped.toUtf8());
```
QByteArray::fromPercentEncoding has wrong behavior too?

SeigneurSerpent commented:
@glassez 
> QByteArray::fromPercentEncoding has wrong behavior too?

Percent-escaping produces US-ASCII character sequences. UTF-8 is backward compatible with US-ASCII, i.e. both have the same binary representation in the range 0x00-0x7F. Invoking `escaped.toUtf8()` produces QByteArray of ASCII symbols, which is consumed by QByteArray::fromPercentEncoding(). The function source code seems to confirm that it does what expected:

src/corelib/tools/qbytearray.cpp
```
static void q_fromPercentEncoding(QByteArray *ba, char percent)
{
    if (ba->isEmpty())
        return;

    char *data = ba->data();
    const char *inputPtr = data;

    int i = 0;
    int len = ba->count();
    int outlen = 0;
    int a, b;
    char c;
    while (i < len) {
        c = inputPtr[i];
        if (c == percent && i + 2 < len) {
            a = inputPtr[++i];
            b = inputPtr[++i];

            if (a >= '0' && a <= '9') a -= '0';
            else if (a >= 'a' && a <= 'f') a = a - 'a' + 10;
            else if (a >= 'A' && a <= 'F') a = a - 'A' + 10;

            if (b >= '0' && b <= '9') b -= '0';
            else if (b >= 'a' && b <= 'f') b  = b - 'a' + 10;
            else if (b >= 'A' && b <= 'F') b  = b - 'A' + 10;

            *data++ = (char)((a << 4) | b);
        } else {
            *data++ = c;
        }

        ++i;
        ++outlen;
    }

    if (outlen != len)
        ba->truncate(outlen);
}

void q_fromPercentEncoding(QByteArray *ba)
{
    q_fromPercentEncoding(ba, '%');
}  

QByteArray QByteArray::fromPercentEncoding(const QByteArray &input, char percent)
{
    if (input.isNull())
        return QByteArray();       // preserve null
    if (input.isEmpty())
        return QByteArray(input.data(), 0);

    QByteArray tmp = input;
    q_fromPercentEncoding(&tmp, percent);
    return tmp;
}
```

glassez commented:
I've considered it (I mean URL decoding). I fully agree with @ngosang - your solution is too complicated. Besides, it is located in the wrong application layer. Later I will present you my suggestions for correcting this (I don't have the free time to do so now).

glassez commented:
@SeigneurSerpent we need to fix this problem at the root. But where is the root? That's right - the root is in the RequestParser class. In short, meaning that we must decode the request on the basis that it may contain any binary data in percent-encoded form. So we need to decode it not in QString but to QByteArray (and change Http::Request::gets to `QMap<QByteArray, QByteArray>`). And the decision how to interpret these data should be taken at the application level (WebApplication, Tracker, etc.).

SeigneurSerpent commented:
@glassez 
Done. 
Your suggestion seems to be the perfect solution to the problem. Thanks!

glassez commented on src/webui/webapplication.cpp:
```diff
@@ -224,7 +224,7 @@ void WebApplication::action_public_images()
 {
     CHECK_URI(0);
-    const QStringMap& gets = request().gets;
+    const QByteArrayMap& gets = request().gets;
```
Did you check whether it breaks the web app or not?

glassez commented on src/base/types.h:
```diff
@@ -42,5 +42,6 @@ enum class ShutdownAction
 };
 
 typedef QMap<QString, QString> QStringMap;
+using QByteArrayMap = QMap<QByteArray, QByteArray>;
```
Maybe `RawStringMap`?

Chocobo1 replied:
shouldn't this be a `typedef`?

glassez replied:
This is the same in c++11:
>A typedef-name can also be introduced by an alias-declaration. The identifier following the using keyword becomes a typedef-name and the optional attribute-specifier-seq following the identifier appertains to that typedef-name. It has the same semantics as if it were introduced by the typedef specifier. In particular, it does not define a new type and it shall not appear in the type-id.

Chocobo1 replied:
>This is the same in c++11:

OK. But from the my view point, it comes down to whether you are going to define a type, which is also stated in what you quote:

>it does not define a new type and ...

In this specific scenario, I interpret the author is defining a type, because `QByteArrayMap` is used non locally.
And another weaker argument, the `typedef` is just above it... mixing like this is weird IMO.

Chocobo1 replied:
http://stackoverflow.com/questions/10747810/what-is-the-difference-between-typedef-and-using-in-c11#comment25929444_10748056
seems `using` syntax is extended because of problems working with template.

glassez commented on src/base/http/requestparser.cpp:
```diff
@@ -123,14 +123,22 @@ bool RequestParser::parseStartingLine(const QString &line)
         // Parse GET parameters
 #ifndef QBT_USES_QT5
-        QListIterator<QPair<QString, QString> > i(url.queryItems());
+        QListIterator<QPair<QByteArray, QByteArray> > i(url.encodedQueryItems());
```
Since we using c++11 now don't insert a space between two (three, four ...) brackets:
```c++
QListIterator<QPair<QByteArray, QByteArray>> i(url.encodedQueryItems());
```

And in other places too.

glassez commented on src/base/http/requestparser.cpp:
```diff
@@ -123,14 +123,22 @@ bool RequestParser::parseStartingLine(const QString &line)
+        QListIterator<QPair<QByteArray, QByteArray> > i(url.encodedQueryItems());
+        while (i.hasNext()) {
+            QPair<QByteArray, QByteArray> pair = i.next();
+            m_request.gets[pair.first] = pair.second;
```
Why don't you do "fromPercentEncoding()" here? Or QUrl doing it itself?

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -32,11 +32,16 @@
 #include "tracker.h"
+#include <boost/asio/ip/address.hpp>
+#if defined(BOOST_ASIO_DYN_LINK) || !defined(BOOST_ASIO_SEPARATE_COMPILATION)
+#include <boost/asio/impl/src.hpp>
```
Get that out of here. If qBittorrent is not compiled you have, it is the problem of your build configuration.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -32,11 +32,16 @@
 #include "base/http/server.h"
 #include "base/utils/string.h"
 #include "tracker.h"
+#include <boost/asio/ip/address.hpp>
```
Move this to other system includes and try to keep include list ordered (e.g. don't mix libtorrent/boost/Qt includes).

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -56,22 +61,160 @@ bool Peer::operator==(const Peer &other) const
 }
 
-QString Peer::uid() const
+const QByteArray& Peer::uid() const
```
To return Qt implicitly shared types as const reference is overkill.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -56,22 +61,160 @@ bool Peer::operator==(const Peer &other) const
 }
 
-libtorrent::entry Peer::toEntry(bool noPeerId) const
+Peer::Peer(QString&& ip, QString&& announce_ip, std::string&& peer_id, int port, QByteArray&& uid)
```
>To return Qt implicitly shared types as const reference is overkill.

as well as use move semantics for it.

glassez commented on src/base/bittorrent/tracker.h:
```diff
@@ -49,21 +50,76 @@ namespace Http
+        const QByteArray& uid() const;
+        // converts to appropriate entry taking into account whether 'clientIp' is a public ip address,  i.e. provides
+        // public ip if 'clientIP' belongs to a public network, otherwise gives connection ip
+        libtorrent::entry toEntryForClient(bool noPeerId, const QString& clientIp) const;
```
Since you use PeerBuilder to build a peer no need to implement Peer as class at all. Simple struct is enough here.
As for this method (toEntryForClient()), it has private meaning and can be simply transformed to module wide function `peerToEntry()`. The main idea is to move all private things to .cpp.

glassez replied:
OK. I did some research. Ideally, we should leave in the header only the following forward declarations (outside BitTorrent namespace):
```c++
struct Peer;
struct TrackerAnnounceRequest;

typedef QHash<QString, Peer> PeerList;
typedef QHash<QString, PeerList> TorrentList;
```

glassez replied:
And then it all makes the PeerBuilder class superfluous!
If you apply my suggestions, you will significantly reduce the amount of code.

glassez commented on src/base/bittorrent/tracker.h:
```diff
@@ -49,21 +50,76 @@ namespace Http
+        static std::string escape(const std::string& str);
+    };
+
+    class PeerBuilder
```
It's pure private thing so move this into .cpp.

glassez commented on src/base/bittorrent/tracker.h:
```diff
@@ -49,21 +50,76 @@ namespace Http
+        // ip retrieved from connection
+        QString m_ip;
+        // ip obtained from announce query
+        QString m_announce_ip;
```
Fix coding style for names. We use camelCase (prepended by `m_` prefix for private fields).

glassez commented on src/base/bittorrent/tracker.h:
```diff
@@ -49,21 +50,76 @@ namespace Http
+        int m_port;
+    };
+
+    struct IpUtil
```
We don't allow pure static classes, so you must use namespaces in such cases.
But since here we have only module private functions, then just place it in an anonymous namespace in the .cpp file.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -56,22 +61,160 @@ bool Peer::operator==(const Peer &other) const
+
+Peer PeerBuilder::build()
+{
+    QByteArray m_uid = m_peer_id.length() == PEER_ID_EXPECTED_LENGTH ? m_peer_id
```
`m_` prefixed names is reserved for private class fields.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -56,22 +61,160 @@ bool Peer::operator==(const Peer &other) const
+}
+
+
+boost::asio::ip::address_v4 IpUtil::toAddress(unsigned char a, unsigned char b, unsigned char c, unsigned char d)
```
Please use shortcuts for unsigned types: `uchar`, `ulong` etc.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -56,22 +61,160 @@ bool Peer::operator==(const Peer &other) const
+    if (ip.is_v6()) {
+        address_v6 ipv6 = ip.to_v6();
+        auto addr = ipv6.to_bytes();
+        return (addr[0] >> 1 == 0xFC >> 1)  // FC00::/7   Unique local
```
Please enclose subexpressions for more readability:
```c++
return ((addr[0] >> 1) == (0xFC >> 1))
```

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -56,22 +61,160 @@ bool Peer::operator==(const Peer &other) const
+                && isNonRoutable(toAddress(addr[2], addr[3], addr[4], addr[5])))
+            // 64:ff9b::/96
+            || (addr[0] == 0 && addr[1] == 0x64 && addr[2] == 0xFF && addr[3] == 0x9B
+                && std::all_of(&addr[4], &addr[12], [](char a) {return a == 0;})
```
Fix coding style for one line function:
```c++
{ return a == 0; }
```

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -56,22 +61,160 @@ bool Peer::operator==(const Peer &other) const
+{
+    using namespace boost::asio::ip;
+
+    address client_ip, ip, announce_ip;
```
camelCase

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -56,22 +61,160 @@ bool Peer::operator==(const Peer &other) const
+        return m_ip;
+
+    // if client has public ip, return public ip
+    if(!IpUtil::isNonRoutable(client_ip)) {
```
Add space after `if` (coding style issue)
And below too

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -157,7 +305,8 @@ void Tracker::respondToAnnounceRequest()
         return;
     }
-    annonceReq.peer.peerId = gets.value("peer_id");
+    peerBuilder.setPeerId( gets.value("peer_id"));
```
Remove space after `(`.

glassez commented on src/base/bittorrent/tracker.h:
```diff
@@ -33,6 +33,7 @@
 #define BITTORRENT_TRACKER_H
 
 #include <QHash>
+#include "boost/asio/ip/address.hpp"
```
Use brackets for system includes.

SeigneurSerpent commented:
@glassez 
```
@@ -224,7 +224,7 @@ void WebApplication::action_public_images()
 void WebApplication::action_query_torrents()
 {
     CHECK_URI(0);
-    const QStringMap& gets = request().gets;
+    const QByteArrayMap& gets = request().gets;
```
> Did you check whether it breaks the web app or not?

I completely forgot to test it with Qt4, so I added some conversion code.
Both Qt5 and Qt4 have QString(QByteArray) constructor, but only Qt5 treats the array as Utf-8 sequence.  In Qt5 everything should work like magic because of QString(QByteArray) constructor and implicit construction from QByteArray to QString. As far as Qt4 is concerned, I added explicit conversion from QByteArray to QString.
Request.gets is used in 3 WebApplication functions: action_query_torrents, action_sync_maindata, action_sync_torrent_peers. But these functions appear not to be used by anything else. So I tested each case using handcrafted custom conversion tests.

```
@@ -123,14 +123,22 @@ bool RequestParser::parseStartingLine(const QString &line)
 
         // Parse GET parameters
 #ifndef QBT_USES_QT5
-        QListIterator<QPair<QString, QString> > i(url.queryItems());
+        QListIterator<QPair<QByteArray, QByteArray> > i(url.encodedQueryItems());
+        while (i.hasNext()) {
+            QPair<QByteArray, QByteArray> pair = i.next();
+            m_request.gets[pair.first] = pair.second;
```
> Why don't you do "fromPercentEncoding()" here? Or QUrl doing it itself?

Fixed.


```
 #include "base/preferences.h"
 #include "base/http/server.h"
 #include "base/utils/string.h"
 #include "tracker.h"
+#include <boost/asio/ip/address.hpp>
+#if defined(BOOST_ASIO_DYN_LINK) || !defined(BOOST_ASIO_SEPARATE_COMPILATION)
+#include <boost/asio/impl/src.hpp>
```

> Get that out of here. If qBittorrent is not compiled you have, it is the problem of your build configuration.

I guess I should leave only `defined(BOOST_ASIO_DYN_LINK)`


All remaining issues have been fixed.

SeigneurSerpent commented:
@Chocobo1 
```
@@ -42,5 +42,6 @@ enum class ShutdownAction
 };
 
 typedef QMap<QString, QString> QStringMap;
+using QByteArrayMap = QMap<QByteArray, QByteArray>;
```

>     This is the same in c++11:
> 
> OK. But from the my view point, it comes down to whether you are going to define a type, which is also stated in what you quote:
> 
>     it does not define a new type and ...
> 
> In this specific scenario, I interpret the author is defining a type, because QByteArrayMap is used non locally.
> And another weaker argument, the typedef is just above it... mixing like this is weird IMO.
> 
> 

Done

glassez commented:
>All remaining issues have been fixed.

You will hate to hear this, but I say no. I asked you to simplify your previous code, you made it difficult. I was told that my proposed simplification will reduce the amount of code and you just put the body of some functions in one line with braces in violation of our coding style rules. By the way, you also did not fix other coding style issues (or you just added the new).
Do you still have previous version of code?

SeigneurSerpent commented:
@glassez 
> Do you still have previous version of code?

Yes I have. I guess rebasing was a terrible idea, so I'll do it only after a complete successful review. I restored the branch and created a separate commit. Here is the proposed fix:

[https://github.com/qbittorrent/qBittorrent/pull/4984/commits/69f621df0e080086008d69dfc0b3606d60b7c497]()

Apparently I misunderstood some of your suggestions. So let me comment on some seemingly very strange changes I've made.
I removed all getters from Peer class when transformed it to struct. Also I renamed toEntryForClient() to peerToEntry() and changed its signature to `libtorrent::entry peerToEntry(const Peer& peer, bool noPeerId, const QString& clientIp)` following the comment:

 ```
         bool operator!=(const Peer &other) const;
         bool operator==(const Peer &other) const;
-        QString uid() const;
-        libtorrent::entry toEntry(bool noPeerId) const;
+        const QByteArray& uid() const;
+        // converts to appropriate entry taking into account whether 'clientIp' is a public ip address,  i.e. provides
+        // public ip if 'clientIP' belongs to a public network, otherwise gives connection ip
+        libtorrent::entry toEntryForClient(bool noPeerId, const QString& clientIp) const;
```

> Since you use PeerBuilder to build a peer no need to implement Peer as class at all. Simple struct is enough here.
> As for this method (toEntryForClient()), it has private meaning and can be simply transformed to module wide function peerToEntry(). The main idea is to move all private things to .cpp.

I removed `escape()` function declaration from Perr class and transferred it to unnamed namespace in tracker.cpp following the comment:
```
+
+        // ip retrieved from connection
+        QString m_ip;
+        // ip obtained from announce query
+        QString m_announce_ip;
+        std::string m_peer_id;
+        int m_port;
+        QByteArray m_uid;
+
+        // if 'clientIP' belongs to a public network, returns public address found among connection and announce ip adrresses
+        const QString& getIpForClient(const QString& clientIP) const;
+        // percent-escapes given string
+        static std::string escape(const std::string& str);
+    };
+
+    class PeerBuilder
```

> It's pure private thing so move this into .cpp.

I guess I didn't quite get what you mean by moving pure private thing to .cpp. I would appreciate if you could clarify this matter.

I deleted the PeerBuilder class and put the code generating peer uid in-place, the only thing it did useful, after reading the following comment:
 
```
         bool operator!=(const Peer &other) const;
         bool operator==(const Peer &other) const;
-        QString uid() const;
-        libtorrent::entry toEntry(bool noPeerId) const;
+        const QByteArray& uid() const;
+        // converts to appropriate entry taking into account whether 'clientIp' is a public ip address,  i.e. provides
+        // public ip if 'clientIP' belongs to a public network, otherwise gives connection ip
+        libtorrent::entry toEntryForClient(bool noPeerId, const QString& clientIp) const;
```
> And then it all makes the PeerBuilder class superfluous!
> If you apply my suggestions, you will significantly reduce the amount of code.

Anyway I greatly appreciate any comments or suggestions. Thanks!

glassez commented:
Sorry, I went a little crazy. It's not so bad. My frustration has caused most of the Http::* code. Tomorrow I will comment in detail on this version of the code (not necessary yet to change anything).

glassez commented on src/base/http/requestparser.cpp:
```diff
@@ -123,14 +123,24 @@ bool RequestParser::parseStartingLine(const QString &line)
+
+            m_request.gets[key] = value;
         }
+#endif
```
Now `requestparser.cpp` changes look good.

glassez commented on src/base/http/types.h:
```diff
@@ -64,16 +64,19 @@ namespace Http
         QByteArray data; // File data
     };
 
-    struct Request
```
Just create Request struct with `RawStringMap gets`. As there can be any binary data according to the standard, it is fully consistent with the model (no need to complicate it). As I said, parsing of these data is the task of the application. In our case, we can consider each Action (eg. `action_query_torrents()`) as a separate application, so we should parse GET params in these actions.

glassez replied:
More comments will be tomorrow.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -29,15 +29,23 @@
 #include "tracker.h"
 
+
+
```
Fix coding style.
Only one blank line allowed.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -47,31 +55,126 @@ using namespace BitTorrent;
-QString Peer::uid() const
-{
-    return ip + ":" + QString::number(port);
+namespace {
```
Fix coding style.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -47,31 +55,126 @@ using namespace BitTorrent;
+    }
+
+    boost::asio::ip::address_v4 toAddress(uchar a, uchar b, uchar c, uchar d)
+    { return boost::asio::ip::address_v4((a << 24) + (b << 16) + (c << 8) + d); }
```
Fix coding style.
And in similar cases too.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -47,31 +55,126 @@ using namespace BitTorrent;
+                    && isNonRoutable(toAddress(addr[2], addr[3], addr[4], addr[5])))
+                // 64:ff9b::/96
+                || (addr[0] == 0 && addr[1] == 0x64 && addr[2] == 0xFF && addr[3] == 0x9B
+                    && std::all_of(&addr[4], &addr[12], [](char a) {return a == 0;})
```
Fix coding style for lambda body.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -47,31 +55,126 @@ using namespace BitTorrent;
+                || (addr[0] == 0x20 && addr[1] == 0x02
+                    && isNonRoutable(toAddress(addr[2], addr[3], addr[4], addr[5])))
+                // 64:ff9b::/96
+                || (addr[0] == 0 && addr[1] == 0x64 && addr[2] == 0xFF && addr[3] == 0x9B
```
Although it does not violate our rules (we just don't have anything for this), but it is better to enclose the subexpressions in parentheses. This improves readability and can save you from random errors if you mix up the precedence of operators.

P.S. My "general" comments (like this one) relate not only to the linked line, but to all code.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -47,31 +55,126 @@ using namespace BitTorrent;
+    return m_ip;
 }
 
+
```
Redundant blank line.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -157,7 +263,15 @@ void Tracker::respondToAnnounceRequest()
+    const QByteArray peerId = gets.value("peer_id");
+    annonceReq.peer.m_peerId = std::string(peerId.constData(), peerId.length());
+
+    if(annonceReq.peer.m_peerId.length() == 20)
```
Fix coding style.

glassez commented on src/base/bittorrent/tracker.h:
```diff
@@ -51,28 +51,33 @@ namespace BitTorrent
-        libtorrent::entry toEntry(bool noPeerId) const;
+
+        // ip retrieved from connection
+        QString m_ip;
```
Fix coding style.
`m_` prefix is used for **private** fields only.

Also keep the fields together and methods (including operators) together.

glassez commented on src/base/bittorrent/tracker.h:
```diff
@@ -51,28 +51,33 @@ namespace BitTorrent
 {
     struct Peer
```
>I guess I didn't quite get what you mean by moving pure private thing to .cpp. I would appreciate if you could clarify this matter.

Peer and TrackerAnnounceRequest structs are used internally only so, as I said before, move its declarations to .cpp (not use anonymous namespace for it)  and use forward declarations here (outside BitTorrent namespace):

```c++
struct Peer;
struct TrackerAnnounceRequest;

typedef QHash<QString, Peer> PeerList;
typedef QHash<QString, PeerList> TorrentList;

namespace BitTorrent
{
```

Also move Peer methods definitions in struct itself (no need to separate it since we have it in .cpp).

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -29,15 +29,23 @@
 
+#include <boost/asio/ip/address.hpp>
+#if defined(BOOST_ASIO_DYN_LINK)
+#include <boost/asio/impl/src.hpp>
```
I mean you shouldn't include `<boost/asio/impl/src.hpp>` here. If you can't build qBittorrent without this, you've incorrectly set up your build environment.

SeigneurSerpent replied:
@glassez 
The problem turned out to be that some boost asio dynamic libraries could be compiled without rvalue reference move support. All Travis-ci comilation attempts would fail without this inclusion. So I disabled the generation of moves. Thanks for pointing it out.

glassez commented:
@SeigneurSerpent I finished to comment. Conclusion:
1. `requestparser.cpp` is OK.
2. `session.cpp` seems to be ok.
3. `tracker.*` need to be improved (according to my comments).
4. `http/types.h` needs to be fixed (according to my comments).
5. `abstractwebapplication.*` need to be untouched.
6. `webapplication.cpp` needs to be fixed (you should add GET params parsing in affected Actions).

When you create QString from UTF-8 encoded QByteArray you should use QString::fromUtf8() for both Qt4 and Qt5 (to avoid redundant `#ifdef`s).

P.S. Try not to break the coding style rules, when you will be correct this time.

glassez commented on src/webui/webapplication.cpp:
```diff
@@ -214,6 +214,14 @@ void WebApplication::action_public_images()
 
+namespace
+{
+    inline QString toUtf8(const QByteArray& a)
```
fromUtf8
Meaning: we create QString **from** UTF-8 encoded data.
Besides, I don't think this shortcut is so necessary.

SeigneurSerpent commented:
I hope I got everything right this time.

glassez commented on src/webui/webapplication.cpp:
```diff
@@ -224,12 +232,15 @@ void WebApplication::action_public_images()
-        gets["filter"], gets["category"], gets["sort"], gets["reverse"] == "true",
-        gets["limit"].toInt(), gets["offset"].toInt()
-        ), Http::CONTENT_TYPE_JSON);
+    print(btjson::getTorrents( toUtf8(gets["filter"]),
```
`fromUtf8()` only makes sense for the category. All other parameters have been designed to contain either numbers or latin-1 string. But you may leave it.

Also fix coding style in this line.

glassez commented on src/webui/webapplication.cpp:
```diff
@@ -273,7 +284,7 @@ void WebApplication::action_query_propertiesFiles()
 {
     CHECK_URI(0);
-    print(btjson::getSyncMainData(request().gets["rid"].toInt(),
+    print(btjson::getSyncMainData(toUtf8(request().gets["rid"]).toInt(),
```
fromUtf8() is meaningless here.

glassez commented on src/webui/webapplication.cpp:
```diff
@@ -284,8 +295,8 @@ void WebApplication::action_sync_maindata()
-    print(btjson::getSyncTorrentPeersData(request().gets["rid"].toInt(),
-        request().gets["hash"],
+    print(btjson::getSyncTorrentPeersData(toUtf8(request().gets["rid"]).toInt(),
+        toUtf8(request().gets["hash"]),
```
fromUtf8() is meaningless in these 2 lines.

glassez commented on src/base/bittorrent/tracker.h:
```diff
@@ -49,30 +49,11 @@ namespace Http
-        // Extensions
-        bool noPeerId;
-    };
+    struct Peer;
```
I told you to move these declarations outside the BitTorrent namespace. No need to put these **private** name here.

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -46,30 +51,154 @@ static const int ANNOUNCE_INTERVAL = 1800; // 30min
+    }
+
+    // if 'clientIP' belongs to a public network, returns public address found among connection and announce ip adrresses
+    QString getIpForClient(const QString& clientIp) const;
```
Why method body isn't here?

glassez commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -157,7 +289,15 @@ void Tracker::respondToAnnounceRequest()
+    else
+        annonceReq.peer.uid = QString("%1:%2").arg(annonceReq.peer.ip).arg(annonceReq.peer.port).toUtf8();
+
+
```
Redundant blank line.

glassez commented:
There are still a few things for fixing. The rest now looks good.

>The problem turned out to be that some boost asio dynamic libraries could be compiled without rvalue reference move support. All Travis-ci comilation attempts would fail without this inclusion. So I disabled the generation of moves. Thanks for pointing it out.

It also looks suspicious. Namely, the dependence of this behavior from the inclusion of `<boost/asio/impl/src.hpp>`. I'll try to test it later.

glassez commented:
Besides, for some reason I can't find any mention of `BOOST_ASIO_DISABLE_MOVE` in Boost.Asio documentation.

SeigneurSerpent commented:
@glassez 
> It also looks suspicious. Namely, the dependence of this behavior from the inclusion of <boost/asio/impl/src.hpp>. I'll try to test it later.
No need. I replace std::swap with copy assignment and hereby eliminated all rvalue moves.

glassez commented on src/webui/webapplication.cpp:
```diff
@@ -234,13 +226,13 @@ void WebApplication::action_query_torrents()
-                               toUtf8(gets["offset"]).toInt()),
-           Http::CONTENT_TYPE_JSON);
+    print(btjson::getTorrents(gets["filter"],
+                              QString::fromUtf8(gets["category"].constData(), gets["category"].length()),
```
Damn, I wanted to say that everything is OK... Why do you complicate everything? Why are you not doing just `QString::fromUtf8(gets["category"])`?

SeigneurSerpent replied:
As far as I know, a general Utf-8 byte sequence could contain null byte somewhere in the middle. So I decided to be a little cautious by supplying length. Certainly 
a category could never contain contain such a symbol.
I will remove the length.

glassez replied:
I think the Qt developers took that into consideration and QString::fromUtf8() works properly.

evsh replied:
In the correct UTF-8 string null byte means end of string and can not appear as a part of another character, and that is one of the reasons why UTF-8 is so popular.

SeigneurSerpent replied:
Thanks for the insight!

evsh commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -45,32 +47,157 @@ static const int ANNOUNCE_INTERVAL = 1800; // 30min
+    }
+
+    // get rid of compiler size_t to int conversion warning
+    int toInt(size_t size)
```
I find this function useless and dangerous. It is used, as far as I see, for string parsing only. If (most likely) the length of the string is not larger than `std::numeric_limits<int>::max()`  the check in it is useless. If it is larger, than it is useless nevertheless because truncating the string we reschedule problem, and it might occur at a place, little correlated with this function. Should there be a `static_cast` anybody will feel that something can go wrong and easily find those places (and that would silence compiler as well).

SeigneurSerpent replied:
When I was writing this I was concerned about the general case in which a conversion from 64-bit or 32-bit unsigned size_t to a signed 32-bit int could produce a negative number. If the receiving function does not have a proper check, then it could lead to something like undefined behavior. I chose the minimum of the two numbers just to get the program running in this case. Thanks to another your comment, I eliminated all places where the conversion was necessary, hence problem doesn't really matter anymore. Anyway, thanks for the advice.

evsh commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -45,32 +47,157 @@ static const int ANNOUNCE_INTERVAL = 1800; // 30min
-    return uid() == other.uid();
-}
+    // ip retrieved from connection
+    QString ip;
```
On every usage this string is either converted to `std::string` or used in `QString::arg()`. Why then to store it in 2-byte encoding?

SeigneurSerpent replied:
Good point, I've converted ip strings to boost::asio::ip::addresses. It has allowed to remove some redundant conversion code.

evsh commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -45,32 +47,157 @@ static const int ANNOUNCE_INTERVAL = 1800; // 30min
+    // ip retrieved from connection
+    QString ip;
+    // ip obtained from announce query
+    QString announceIp;
```
This one is not used as `QString` at all as far as I can see, and gets converted to `std::string` always.

evsh commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -45,32 +47,153 @@ static const int ANNOUNCE_INTERVAL = 1800; // 30min
+    std::string escape(const QByteArray& str)
+    {
+        // like libtorrent percent-encoding
+        QByteArray encoded = str.toPercentEncoding("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~!()*");
```
Although I haven't tried it myself, Qt5 documentation example for this function shows that characters from the `exclude` parameter are used as addition to the standard set of [a-zA-Z0-9] and "-" / "." / "_" / "~". Thus, I believe, the call can be simplified:
```C++
QByteArray encoded = str.toPercentEncoding("!()*");
```

SeigneurSerpent replied:
Corrected

evsh commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -45,32 +47,153 @@ static const int ANNOUNCE_INTERVAL = 1800; // 30min
+        ip = static_cast<const address>(address::from_string(Utils::String::toStdString(ipStr), ec));
+        announceIp = static_cast<const address>(address::from_string(string(announceIpStr), ec));
+
+        uid = peerId.length() == 20 ? peerId : QString("%1:%2").arg(ipStr).arg(port).toUtf8();
```
The `toUtf8()` call here confuses me. Can a string consisting of an IP and a port number contain anything except ASCII?

glassez replied:
But you know that toUtf8() will produce the correct value here! Or do you fundamentally want to have toLaton1() here?

evsh replied:
I want the code to be clean and to have obvious meaning. `toUtf8()` makes me think that this piece of the code might be a leftover from a refactoring, a part of string might be forgotten or something else. This gives me a feeling that I simply don't understand this fragment. So, yes, I want either a comment or `toLatin1()`.

SeigneurSerpent replied:
Amended

sledgehammer999 commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -45,32 +47,153 @@ static const int ANNOUNCE_INTERVAL = 1800; // 30min
+        ip = static_cast<const address>(address::from_string(Utils::String::toStdString(ipStr), ec));
+        announceIp = static_cast<const address>(address::from_string(string(announceIpStr), ec));
+
+        uid = peerId.length() == 20 ? peerId : QString("%1:%2").arg(ipStr).arg(port).toLatin1();
```
Keep in mind that I am not good with networking details but wouldn't be here to use the `announceIpStr` instead of the `ipStr`. (In case 2 different peers come from inside the same NAT).

sledgehammer999 replied:
Scratch that. I know realize that the NAT's ip would be a public ip, but a peer from behind a NAT would see/use a private ip.

sledgehammer999 commented on src/webui/webapplication.cpp:
```diff
@@ -224,12 +224,15 @@ void WebApplication::action_public_images()
+    const RawStringMap& gets = request().gets;
+
+    print(btjson::getTorrents(gets["filter"],
+                              QString::fromUtf8(gets["category"]),
```
Unless I am mistaken, this is a totally unrelated change, right?

SeigneurSerpent replied:
The PR changes the way GET query parameters are decoded. RequestParser used to convert them directly into QString expecting the percent-encoded byte sequence to be UTF-8. On the other hand, tracker announce url can contain arbitrary percent-encoded 20-byte sequences in info_hash and peer_id parameters that have no valid UTF-8 representation. Therefore RequestParser is change to decode to QByteArray and anything that expects QString should make a second conversion from QByteArray to QString. Qt4 makes implicit conversion to QString using  fromAscii() and Qt5 uses fromUtf8(). That is why explicit QString::fromUtf8() is used for 'category'.

sledgehammer999 commented on src/base/bittorrent/session.cpp:
```diff
@@ -2557,7 +2557,15 @@ void Session::handleListenFailedAlert(libt::listen_failed_alert *p)
+    if (Preferences::instance()->getNetworkAddress().isEmpty()) {
+        libt::session_settings sessionSettings = m_nativeSession->settings();
+        sessionSettings.announce_ip = newIp;
+        m_nativeSession->set_settings(sessionSettings);
```
You need to save the external ip. Currently the session_settings are reapplied if the user changes something in the settings window. Thus, in that case the possibly empty announce_ip will override the external ip that you set here.
Take a look at `Session::setSessionSettings()`.

SeigneurSerpent replied:
Fixed

sledgehammer999 commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -240,7 +370,7 @@ void Tracker::replyWithPeerList(const TrackerAnnounceRequest &annonceReq)
     foreach (const Peer &p, peers) {
         //if (p != annonce_req.peer)
-        peerList.push_back(p.toEntry(annonceReq.noPeerId));
+        peerList.push_back(peerToEntry(p, annonceReq.noPeerId, annonceReq.peer.ip));
```
I am bit confused here. I tracked back where `annonceReq.peer.ip` comes from and I think it is the IP of the client requesting the peer list. So why does `peerToEntry()` potentially return his IP back?

sledgehammer999 replied:
>I am bit confused here. I tracked back where annonceReq.peer.ip comes from and I think it is the IP of the client requesting the peer list. So why does peerToEntry() potentially return his IP back?

Damnit. I messed up reading the code. Ignore The above. I'll come back if I figure it out.

sledgehammer999 commented:
@SeigneurSerpent one final comment. I may be a little bit confused by your changes. But here it goes. It is about the announce ip. From [here](https://wiki.theory.org/BitTorrentSpecification#Tracker_Request_Parameters):
> ip: Optional. The true IP address of the client machine, in dotted quad format or rfc3513 defined hexed IPv6 address. Notes: In general this parameter is not necessary as the address of the client can be determined from the IP address from which the HTTP request came. The parameter is only needed in the case where the IP address that the request came in on is not the IP address of the client. This happens if the client is communicating to the tracker through a proxy (or a transparent web proxy/cache.) It also is necessary when both the client and the tracker are on the same local side of a NAT gateway. The reason for this is that otherwise the tracker would give out the internal (RFC1918) address of the client, which is not routable. Therefore the client must explicitly state its (external, routable) IP address to be given out to external peers. Various trackers treat this parameter differently. **Some only honor it only if the IP address that the request came in on is in RFC1918 space. Others honor it unconditionally, while others ignore it completely.** In case of IPv6 address (e.g.: 2001:db8:1:2::100) it indicates only that client can communicate via IPv6.

So IMO if the peer has set an announce ip, it should take precedence over the IP it connected through AND if it is routable, otherwise use the connected IP as fallback.

PS: If you know what interactive rebase is, then use the `fixup` keyword to merge your commits. If not, leave a comment and I'll give detailed instructions.

sledgehammer999 commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -45,32 +47,153 @@ static const int ANNOUNCE_INTERVAL = 1800; // 30min
+            return ip;
+
+        // if client has public ip, return public ip
+        if (!isNonRoutable(clientIp)) {
```
Why should we care if the currently connected peer has a public IP?

SeigneurSerpent replied:
The purpose is to give external ip to external client and otherwise - connection ip. Otherwise two local clients on the same network behind a NAT router might not be able to share private torrents. If public ip is announced by default(inferred router's external ip and set in Session::handleExternalIPAlert()) and tracker returns external announceIp to local clients, then they will try to connect to the router and fail.

sledgehammer999 commented on src/base/bittorrent/tracker.cpp:
```diff
@@ -45,32 +47,153 @@ static const int ANNOUNCE_INTERVAL = 1800; // 30min
+    const address& getIpForClient(const address& clientIp) const
+    {
+        // if any address is unspecified, return ip
+        if (clientIp.is_unspecified() || ip.is_unspecified() || announceIp.is_unspecified())
```
The same here. Why should we care? The routable test should take of it.

SeigneurSerpent replied:
Fixed

sledgehammer999 commented:
@qbittorrent/qbittorrent-frequent-contributors it seems that @SeigneurSerpent has abandonded this PR(or too busy). Can someone review my comments, so I can prepare a patch after merging this?

SeigneurSerpent commented:
@sledgehammer999 
Sorry for taking so long to reply. I was away from home.

SeigneurSerpent commented:
@sledgehammer999 

If the changes are acceptable I can promptly merge them by interactive rebase

sledgehammer999 commented:
I took a glance at your comments. I'll look the changes more thoroughly tomorrow.
You haven't commented on my https://github.com/qbittorrent/qBittorrent/pull/4984#issuecomment-216602483

SeigneurSerpent commented:
> So IMO if the peer has set an announce ip, it should take precedence over the IP it connected through AND if it is routable, otherwise use the connected IP as fallback.

I've changed somewhat server logic following your suggestion and now server provides external annouceIp to external client, otherwise - connection IP. 
However, you might be suggesting to always provide announceIp if it is routable no matter what client IP is. If that's the case, lets consider the setting of two local clients in the same network sharing private torrent. Current client implementation always announces external Ip in Session::handleExternalIPAlert(). Therefore both clients would announce their router's external IP to the server(which is one of the clients). Torrent server would provide the external IP and the clients would fail to initiate data transfer. Even if both specify their own local IP as annouceIp. On the other hand, current server implementation would provide local IPs to both clients and they would successfully communicate with each other.
From the conceptual point of view, it is certainly right to provide routable announceIp. For the above-mentioned scenario to work, I guess I should introduce a boolean option to turn inferred external ip announcement off.

ngosang commented on src/base/bittorrent/session.cpp:
```diff
@@ -2557,7 +2558,15 @@ void Session::handleListenFailedAlert(libt::listen_failed_alert *p)
     boost::system::error_code ec;
-    Logger::instance()->addMessage(tr("External IP: %1", "e.g. External IP: 192.168.0.1").arg(p->external_address.to_string(ec).c_str()), Log::INFO);
+    m_externalIp = p->external_address.to_string(ec);
+    Logger::instance()->addMessage(tr("External IP: %1", "e.g. External IP: 192.168.0.1").arg(m_externalIp.c_str()), Log::INFO);
```
Minor change: 192.168.0.1 is a LAN IP, change it for a WAN IP e.g. 87.125.203.161

aekkarinnan commented:
6

glassez commented:
@Chocobo1, what do you think about this feature in general?

Chocobo1 commented:
>@Chocobo1, what do you think about this feature in general?

I'm working on something similar (improving the basic tracker), although not as complex as the PR. I'll put up for review when ready.
